### PR TITLE
many: add apparmorprompting interface manager

### PR DESCRIPTION
**Pull request by olivercalder**
This PR is based on the current outstanding prompting PRs, which are aggregated under #14314. Only the last few commits are relevant to the PR, others will be removed as the preceding PRs are merged.

The code in this PR was originally based on the udevmonitor package. Much of this PR should be rewritten and simplified, as there's no reason for this manager to be governed by the `Connect`, `Run`, `Stop` methods and repeatedly restarted via `Ensure`.

-----

Add the `apparmorprompting` interface manager to start and manage interactions between the listener, prompts backend, and rules backend.

This package mediates interactions between the prompting API and these backends, and between the listener and these backends.

This PR is tracked internally by https://warthogs.atlassian.net/browse/SNAPDENG-10089

### CODE REVIEW

**Comment by olivercalder** [comment]
Questions remain: should prompting manager be started by `InterfaceManager.StartUp()` or by `InterfaceManager.Ensure()`? I would prefer the former, but I'm not sure if there are architectural requirements to the contrary. And there's lots of duck typing. @zyga @bboozzoo what do you think?

**Comment by pedronis** [comment]
> Questions remain: should prompting manager be started by InterfaceManager.StartUp() or by InterfaceManager.Ensure()? I would prefer the former, but I'm not sure if there are architectural requirements to the contrary. And there's lots of duck typing. @zyga @bboozzoo what do you think?

given that we restart snapd when the feature flag changes and we run this only if the feature is on I suppose, StartUp should be ok. Ensure would be more appropriate if the feature didn't require restarting snapd. You still need to take care of tear down in Stop though

File: overlord/ifacestate/apparmorprompting/prompting.go [hunk 1]
```diff
@@ -0,0 +1,551 @@
...
+		// XXX: this log leaks information about internal activity
+		logger.Noticef("error while parsing AppArmor permissions: %v", err)
+		response := listener.Response{
+			Allow:      false,
+			Permission: req.Permission,
+		}
+		return requestReply(req, &response)
+	}
+
+	remainingPerms := make([]string, 0, len(permissions))
```
Comment (bboozzoo) [inline overlord/ifacestate/apparmorprompting/prompting.go:199]: I think you only need to take the lock around here or do replies need to be ordered as well?

File: overlord/ifacestate/apparmorprompting/prompting.go [hunk 2]
```diff
@@ -0,0 +1,551 @@
...
+		// Error should not occur, but if it does, responsePermissions are set
+		// to none, leaving it to AppArmor to default deny.
+
+		response := listener.Response{
+			Allow:      true,
+			Permission: responsePermissions,
+		}
+		return requestReply(req, &response)
+	}
+
```
Comment (bboozzoo) [inline overlord/ifacestate/apparmorprompting/prompting.go:238]:
```suggestion

// request not satisfied by any of existing rules, record a prompt for the user
```

File: overlord/ifacestate/apparmorprompting/prompting.go [hunk 3]
```diff
@@ -0,0 +1,551 @@
...
+	defer m.lock.RUnlock()
+	return m.prompts.PromptWithID(userID, promptID)
+}
+
+// HandleReply checks that the given reply contents are valid, satisfies the
+// original request, and does not conflict with any existing rules (if lifespan
+// is not "single"). If all of these are true, sends a reply for the prompt with
+// the given ID, and both creates a new rule and checks any outstanding prompts
+// against it, if the lifespan is not "single".
+func (m *InterfacesRequestsManager) HandleReply(userID uint32, promptID prompting.IDType, constraints *prompting.Constraints, outcome prompting.OutcomeType, lifespan prompting.LifespanType, duration string) (satisfiedPromptIDs []prompting.IDType, retErr error) {
```
Comment (bboozzoo) [inline overlord/ifacestate/apparmorprompting/prompting.go:344]: this does not appear to be called?

**Review by bboozzoo — COMMENTED** [review]
Some initial comments. The unit tests are failing ATM

File: sandbox/apparmor/notify/listener/listener.go [hunk 4]
```diff
@@ -323,18 +298,20 @@ func (l *Listener) Run() error {
 	// is called) return and close the tomb's dead channel, as it is the last
 	// and only tracked goroutine.
 	l.tomb.Go(func() error {
-		var err error
 		for {
-			err = l.tomb.Err()
-			if err != tomb.ErrStillAlive {
+			if l.tomb.Err() != tomb.ErrStillAlive {
 				break
```
Comment (pedronis) [inline sandbox/apparmor/notify/listener/listener.go:303]: shouldn't we also log this error or is it logged at another level?

File: sandbox/apparmor/notify/listener/listener.go [hunk 5]
```diff
@@ -323,18 +298,20 @@ func (l *Listener) Run() error {
...
 			}
-			err = l.runOnce()
+			err := l.runOnce()
 			if err != nil {
-				break
+				if exitOnError {
+					return err
+				} else {
+					logger.Noticef("error in listener run loop: %v", err)
```
Comment (pedronis) [inline sandbox/apparmor/notify/listener/listener.go:310]: can the code really continue at this point?

File: overlord/ifacestate/ifacemgr.go [hunk 6]
```diff
@@ -453,8 +507,9 @@ func (m *InterfaceManager) DisableUDevMonitor() {
 }
 
 var (
-	udevInitRetryTimeout = time.Minute * 5
-	createUDevMonitor    = udevmonitor.New
+	udevInitRetryTimeout            = time.Minute * 5
+	createUDevMonitor               = udevmonitor.New
+	createInterfacesRequestsManager = apparmorprompting.New
```
Comment (pedronis) [inline overlord/ifacestate/ifacemgr.go:512]: probably a good idea if the two side match?

File: overlord/ifacestate/apparmorprompting/prompting.go (hunk 3 above)
Comment (pedronis) [reply overlord/ifacestate/apparmorprompting/prompting.go:344]: it will be called by daemon I think

**Review by pedronis — COMMENTED** [review]
I'm slightly worried about the changes in listener as it took us a while to land it. Also apparmorprompting tests are not "go test -race" clean atm

File: overlord/ifacestate/apparmorprompting/prompting.go [hunk 7]
```diff
@@ -0,0 +1,551 @@
...
+}
+
+func (m *InterfacesRequestsManager) applyRuleToOutstandingPrompts(rule *requestrules.Rule) (satisfiedPromptIDs []prompting.IDType, err error) {
+	metadata := &prompting.Metadata{
+		User:      rule.User,
+		Snap:      rule.Snap,
+		Interface: rule.Interface,
+	}
+	satisfiedPromptIDs, err = m.prompts.HandleNewRule(metadata, rule.Constraints, rule.Outcome)
+	return satisfiedPromptIDs, err
```
Comment (pedronis) [inline overlord/ifacestate/apparmorprompting/prompting.go:438]: given that the callers are always ignoring errors maybe we can do it directly here? it avoid if and repeated code above

File: overlord/ifacestate/apparmorprompting/prompting_test.go [hunk 8]
```diff
@@ -217,7 +225,12 @@ func (s *apparmorpromptingSuite) TestHandleListenerErrors(c *C) {
 	}
 	reqChan <- req
 	time.Sleep(10 * time.Millisecond)
-	c.Check(fmt.Errorf("%#v", logbuf.String()), ErrorMatches, ".*Error while handling request: cannot get abstract permissions from empty AppArmor permissions.*")
+	logger.WithLoggerLock(func() {
+		c.Check(logbuf.String(), testutil.Contains,
+			" WARNING: too many outstanding prompts for user 1000; auto-denying new one\n")
+	})
+
+	c.Assert(mgr.Stop(), IsNil)
```
Comment (pedronis) [inline overlord/ifacestate/apparmorprompting/prompting_test.go:233]: shouldn't we defer these as early as reasonable instead?

File: overlord/ifacestate/apparmorprompting/prompting_test.go [hunk 9]
```diff
@@ -170,23 +171,27 @@ func (s *apparmorpromptingSuite) TestHandleListenerErrors(c *C) {
...
-	defer restore()
 	req := &listener.Request{
 		// Most fields don't matter here
 		Permission: notify.FilePermission(0),
 	}
 	reqChan <- req
 	time.Sleep(10 * time.Millisecond)
-	c.Check(fmt.Errorf("%#v", logbuf.String()), ErrorMatches, ".*error while parsing AppArmor permissions: cannot get abstract permissions from empty AppArmor permissions.*")
+	logger.WithLoggerLock(func() {
+		c.Check(logbuf.String(), testutil.Contains,
```
Comment (olivercalder) [inline overlord/ifacestate/apparmorprompting/prompting_test.go:192]: I didn't know about testutil.Contains :sweat_smile: thanks!

File: overlord/ifacestate/apparmorprompting/prompting_test.go [hunk 10]
```diff
@@ -264,8 +281,20 @@ func (s *apparmorpromptingSuite) prepManagerWithSinglePrompt(c *C, reqChan chan
 		Class:      notify.AA_CLASS_FILE,
 		Permission: notify.AA_MAY_READ,
 	}
+	// push a request
 	reqChan <- req
 
+	// which should generate a notice
+	s.st.Lock()
+	ctx, cancel := context.WithTimeout(context.Background(), time.Minute)
```
Comment (olivercalder) [inline overlord/ifacestate/apparmorprompting/prompting_test.go:289]: timeout can be shorter I think, more like 10s on the high end. Usually sub-1s is actually fine

File: overlord/ifacestate/apparmorprompting/prompting_test.go [hunk 11]
```diff
@@ -256,6 +271,8 @@ func (s *apparmorpromptingSuite) prepManagerWithSinglePrompt(c *C, reqChan chan
 
 	path := "/home/test/foo"
 
+	whenSent := time.Now()
```
Comment (olivercalder) [inline overlord/ifacestate/apparmorprompting/prompting_test.go:274]: I spent a while yesterday trying to make it easier to check new notices by putting `lastNoticeTimestamp` into `apparmorpromptingSuite` and adding some check functions akin to those in `requestrules`, but ended up throwing it away. I think this is a good simple approach. And in general this package doesn't care that much when notices are recorded, that's more a problem for `requestprompts` and `requestrules`, though it's nice to check when we can, especially when interesting synchronization is required between the two backends.

**Review by olivercalder — COMMENTED** [review]
Thanks for the changes!

File: sandbox/apparmor/notify/listener/listener.go (hunk 5 above)
Comment (olivercalder) [reply sandbox/apparmor/notify/listener/listener.go:311]: Yes, that was the idea behind logging the error instead of returning it. Since all the other internal errors are now logged instead of returned from within tomb-tracked goroutines, the tomb is only ever killed by `listener.Close()`, so the listener will happy carry on even if an error occurs.

File: sandbox/apparmor/notify/listener/listener.go (hunk 4 above)
Comment (olivercalder) [reply sandbox/apparmor/notify/listener/listener.go:304]: Since we've replaced the error returns from tracked goroutines with logs instead, the only error should be the listener being closed. But I don't see that logged elsewhere, so you're right, would be good to log here.

File: sandbox/apparmor/notify/listener/listener.go (hunk 5 above)
Comment (pedronis) [reply sandbox/apparmor/notify/listener/listener.go:313]: s/listener/prompting listener/ ?

File: overlord/ifacestate/apparmorprompting/prompting_test.go [hunk 12]
```diff
@@ -0,0 +1,731 @@
...
+	c.Assert(mgr.Stop(), IsNil)
+}
+
+func (s *apparmorpromptingSuite) TestNewRuleHandlesExistingPrompt(c *C) {
+	//reqChan, replyChan, restore := apparmorprompting.MockListener()
+	//defer restore()
+
+	//mgr, err := apparmorprompting.New(s.st)
+	//c.Assert(err, IsNil)
+	c.Fatalf("TODO")
```
Comment (pedronis) [inline overlord/ifacestate/apparmorprompting/prompting_test.go:649]: we cannot have Fatalf tests on master, you need to use Skip

File: overlord/ifacestate/ifacemgr.go [hunk 13]
```diff
@@ -137,6 +141,20 @@ func Manager(s *state.State, hookManager *hookstate.HookManager, runner *state.T
 	return m, nil
 }
 
+// AppArmorPromptingRunning returns true if prompting is running.
+func (m *InterfaceManager) AppArmorPromptingRunning() bool {
+	return useAppArmorPrompting(m)
```
Comment (pedronis) [inline overlord/ifacestate/ifacemgr.go:146]: this is a bit of a weird one because the first time this is called we need the state locked and after that not strictly, my preference would be to expose only the cached value instead maybe?

File: overlord/ifacestate/ifacemgr.go [hunk 14]
```diff
@@ -174,6 +192,17 @@ func (m *InterfaceManager) StartUp() error {
...
 		return err
 	}
+	if useAppArmorPrompting(m) {
+		if err := m.initInterfacesRequestsManager(); err != nil {
+			// TODO: if this fails, set useAppArmorPromptingValue to false ?
+			// If this is done before profilesNeedRegeneration, then profiles will only
+			// be regenerated if prompting is newly supported and the backends were
+			// successfully created.
+			// TODO: also set "apparmor-prompting" flag to false?
+			// XXX: should this log an error, rather than returning error?
```
Comment (pedronis) [inline overlord/ifacestate/ifacemgr.go:202]: I think yes, that this should only log as long as continuing does reasonable things

File: overlord/ifacestate/apparmorprompting/prompting.go
Comment (pedronis) [inline overlord/ifacestate/apparmorprompting/prompting.go:1]: go test -race seems not clean again here, please try to fix

File: overlord/ifacestate/apparmorprompting/prompting.go [hunk 15]
```diff
@@ -0,0 +1,578 @@
...
+			logger.Noticef("InterfacesRequestsManager tomb is dying, disconnecting")
+			break run_loop
+		}
+	}
+	return m.disconnect()
+}
+
+func (m *InterfacesRequestsManager) handleListenerReq(req *listener.Request) error {
+	userID := uint32(req.SubjectUID)
+	// TODO: immediately deny if req.SubjectUID() == 0 (root)
```
Comment (pedronis) [inline overlord/ifacestate/apparmorprompting/prompting.go:178]: please implement this

File: overlord/ifacestate/apparmorprompting/prompting.go [hunk 16]
```diff
@@ -0,0 +1,578 @@
...
+		} else {
+			if !errors.Is(err, requestrules.ErrNoMatchingRule) {
+				logger.Noticef("error while checking request against existing rules: %v", err)
+			}
+			// No matching rule found
+			remainingPerms = append(remainingPerms, perm)
+		}
+	}
+	if matchedDenyRule {
+		// XXX: this debug log leaks information about internal activity
```
Comment (pedronis) [inline overlord/ifacestate/apparmorprompting/prompting.go:227]: can you explain the what is the issue here XXX? isn't that the nature of debug logs?

File: overlord/ifacestate/apparmorprompting/prompting.go [hunk 17]
```diff
@@ -0,0 +1,578 @@
...
+		logger.Debugf("request denied by existing rule: %+v", req)
+
+		// At least some permissions were allowed, but others were denied.
```
Comment (pedronis) [inline overlord/ifacestate/apparmorprompting/prompting.go:230]: shouldn't we check that len(satisfiedPerms) > 0

File: overlord/ifacestate/apparmorprompting/prompting.go [hunk 18]
```diff
@@ -0,0 +1,578 @@
...
+		// Respond with this information by allowing all the permissions which
+		// were explicitly allowed, and let the kernel auto-deny the denied
+		// permissions, and any which not matched by allow rules.
+		responsePermissions, _ := prompting.AbstractPermissionsToAppArmorPermissions(iface, satisfiedPerms)
+		// Error should not occur, but if it does, responsePermissions are set
+		// to none, leaving it to AppArmor to default deny.
+		response := listener.Response{
+			Allow:      true,
+			Permission: responsePermissions,
```
Comment (pedronis) [inline overlord/ifacestate/apparmorprompting/prompting.go:239]: what if permission is empty? is that equivalent to deny? shouldn't we proceed to make a prompt then?

File: overlord/ifacestate/apparmorprompting/prompting.go [hunk 19]
```diff
@@ -0,0 +1,578 @@
...
+	if err != nil {
+		// Should not occur, as outcome and constraints have already been
+		// validated. However, it's possible an error could occur if the prompt
+		// DB was already closed. This should be an internal error. However, we
+		// can't un-send any reply which might have been sent, and this should
+		// only fail if the prompting system is shutting down, so don't actually
+		// return an error.
+
+		// XXX: this log leaks information about internal activity
+		logger.Noticef("error when handling new rule as a result of reply: %v", err)
```
Comment (pedronis) [inline overlord/ifacestate/apparmorprompting/prompting.go:452]: as we are doing this in all places we are calling  m.applyRuleToOutstandingPrompts maybe we should move it and have it not return an error?

File: overlord/ifacestate/apparmorprompting/prompting.go [hunk 20]
```diff
@@ -0,0 +1,578 @@
...
+	var newRule *requestrules.Rule
+	if lifespan != prompting.LifespanSingle {
+		// Check that adding the rule doesn't conflict with other rules
+		newRule, err = m.rules.AddRule(userID, prompt.Snap, prompt.Interface, constraints, outcome, lifespan, duration)
+		if err != nil {
+			// Rule conflicts with existing rule (at least one identical pattern
+			// variant and permission). This should be considered a bad reply,
+			// since the user should only be prompted for permissions and paths
+			// which are not already covered.
+			return nil, err
```
Comment (pedronis) [inline overlord/ifacestate/apparmorprompting/prompting.go:421]: we always end up here in tests according to coverage, what am I missing? that we don't support other Lifespans in rules atm?

File: overlord/ifacestate/apparmorprompting/prompting.go [hunk 21]
```diff
@@ -0,0 +1,578 @@
...
+	prompt, retErr = m.prompts.Reply(userID, promptID, outcome)
+	if retErr != nil {
+		// Error should not occur unless the listener has closed
+		return nil, retErr
+	}
+
+	if lifespan == prompting.LifespanSingle {
+		return []prompting.IDType{}, nil
+	}
+
```
Comment (pedronis) [inline overlord/ifacestate/apparmorprompting/prompting.go:440]: basically we cannot reach here atm, which is fine but maybe we need to have a comment here about that

**Review by pedronis — CHANGES_REQUESTED** [review]
did a pass, some comments, main thing I'm not sure about the logic in handleListenerReq in the corner case when we don't satisfy anything, maybe I'm missing something but at a minimum some of the comments there are slightly misleading

**Comment by pedronis** [comment]
@olivercalder @bboozzoo also go test -race is failing again in apparmorprompting

File: overlord/ifacestate/apparmorprompting/prompting_test.go (hunk 8 above)
Comment (olivercalder) [reply overlord/ifacestate/apparmorprompting/prompting_test.go:259]: When these are deferred the tests explode due to the way the listener is mocked, unfortunately

File: overlord/ifacestate/apparmorprompting/prompting.go (hunk 16 above)
Comment (olivercalder) [reply overlord/ifacestate/apparmorprompting/prompting.go:234]: The concern is about logging the private application usage activity of users in the logs. But this is a debug message, so it's probably okay.

File: overlord/ifacestate/apparmorprompting/prompting.go (hunk 20 above)
Comment (olivercalder) [reply overlord/ifacestate/apparmorprompting/prompting.go:428]: We don't support rules with `LifespanSingle` (hence the checks when we're replying to a prompt, for which a reply can have `LifespanSingle`), but the tests should add `LifespanForever` and `LifespanTimespan` rules. I think all should be exercised now?

File: overlord/ifacestate/apparmorprompting/prompting.go (hunk 21 above)
Comment (olivercalder) [reply overlord/ifacestate/apparmorprompting/prompting.go:447]: Ahh, yes tests cover this now

File: overlord/ifacestate/apparmorprompting/prompting.go (hunk 7 above)
Comment (olivercalder) [reply overlord/ifacestate/apparmorprompting/prompting.go:472]: Really good idea, thank you!

File: overlord/ifacestate/apparmorprompting/prompting.go (hunk 16 above)
Comment (olivercalder) [reply overlord/ifacestate/apparmorprompting/prompting.go:234]: I've removed all these comments and the actual request details from the messages, now we'll just get the errors.

File: overlord/ifacestate/ifacemgr.go (hunk 13 above)
Comment (olivercalder) [reply overlord/ifacestate/ifacemgr.go:146]: The problem is how to ensure that the cached value has actually been computed. If we're fairly sure that the call to `useAppArmorPrompting` in `StartUp` will be the first time it can be called, then I'm comfortable having this method just export the value, since we don't call this method from within this package.

In fact, we don't call this method from anywhere yet, it is only used from the API. So perhaps just exporting the value is perfectly safe.

We could do something safer by adding a channel which is closed when the `sync.Once` in `useAppArmorPrompting` sets the value, but then there's a chance of blocking during the `AppArmorPromptingRunning` call if for some reason it happens before `StartUp`.

Another alternative would be to do something like this:
```go
func (m *InterfaceManager) AppArmorPromptingRunning() bool {
	m.useAppArmorPromptingChecker.Do(func() {
		m.state.Lock()
		defer m.state.Unlock()
		useAppArmorPrompting(m)
	}
	return m.useAppArmorPromptingValue
}
```

So I see a few options:
1. Just return the current value, even if it's uninitialized
2. Add a channel which `AppArmorPromptingRunning` waits to close before returning the current value
3. Make `AppArmorPromptingRunning` call a closure which locks state and then calls `m.useAppArmorPrompting`

File: overlord/ifacestate/apparmorprompting/prompting.go (hunk 17 above)
Comment (olivercalder) [reply overlord/ifacestate/apparmorprompting/prompting.go:233]: We don't mind whether `len(satisfiedPerms) > 0` or not, the comment is misleading. I'll adjust it to explicitly mention that there be no allowed permissions.

It's increasingly clear to me that the `listener.Response` system (which has fields `Allow bool` and `Permission any`) is unnecessary and confusing, since the `Deny` value basically throws out the `Permission` and just puts all originally requested permissions into the deny field of the message to the kernel $^1$ . A better system would make this more explicit: reply with any permissions you want to allow, any other permissions will be denied. Then the listener can AND the allowed permissions with the originally requested permissions, and send that back to the kernel.

$.^1$ there's a bit of nuance here, the kernel actually sends both an allow and a deny field: the allow field is permissions allowed by the profile, the deny field is permissions matched by rules with the prompt prefix. So we leave the allow field as-is, and treat the deny field as the requested permissions.

File: overlord/ifacestate/apparmorprompting/prompting.go (hunk 18 above)
Comment (olivercalder) [reply overlord/ifacestate/apparmorprompting/prompting.go:242]: If there's a matching deny rule (with higher precedence than any allow rule) for any permission, we immediately send back a denial for any permissions which aren't explicitly allowed. If we happen to have rules allowing some of the requested permissions, we might as well put those permissions in the allow field we're sending back to the kernel, but any explicitly denied permissions or permissions without a matching rule will be put in the deny field going back to the kernel.

File: overlord/ifacestate/apparmorprompting/export_test.go [hunk 22]
```diff
@@ -0,0 +1,122 @@
...
+
+import (
+	"github.com/snapcore/snapd/interfaces/prompting/requestprompts"
+	"github.com/snapcore/snapd/interfaces/prompting/requestrules"
+	"github.com/snapcore/snapd/sandbox/apparmor/notify/listener"
+	"github.com/snapcore/snapd/testutil"
+)
+
+func MockListenerRegister(f func() (*listener.Listener, error)) (restore func()) {
+	restore = testutil.Backup(&listenerRegister)
```
Comment (bboozzoo) [inline overlord/ifacestate/apparmorprompting/export_test.go:30]: these can use testutil.Mock

File: overlord/ifacestate/apparmorprompting/prompting.go (hunk 18 above)
Comment (bboozzoo) [reply overlord/ifacestate/apparmorprompting/prompting.go:242]: in listener we'll hit this snippet:
```
	perms, ok := response.Permission.(notify.FilePermission)
	...
		// allow permissions which kernel initially allowed, along with those
		// which the were initially denied but the user explicitly allowed.
		resp.Allow = msg.Allow | (uint32(perms) & msg.Deny)
		resp.Deny = 0
		resp.Error = 0
```

I'm wondering, shouldn't deny be filled with `~uint32(perms) & msg.Deny` instead of 0?

File: overlord/ifacestate/export_test.go [hunk 23]
```diff
@@ -128,6 +129,30 @@ func MockCreateUDevMonitor(new func(udevmonitor.DeviceAddedFunc, udevmonitor.Dev
 	}
 }
 
+func MockCreateInterfacesRequestsManager(new func(s *state.State) (*apparmorprompting.InterfacesRequestsManager, error)) (restore func()) {
+	old := createInterfacesRequestsManager
```
Comment (bboozzoo) [inline overlord/ifacestate/export_test.go:133]: testutil.Mock

File: overlord/ifacestate/ifacemgr.go (hunk 13 above)
Comment (bboozzoo) [reply overlord/ifacestate/ifacemgr.go:146]: AFAIU first time this will be called in StartUp() of ifacemgr. I don't think anything else may need it.

File: overlord/ifacestate/apparmorprompting/export_test.go [hunk 24]
```diff
@@ -0,0 +1,122 @@
...
+}
+
+func MockListenerRun(f func(l *listener.Listener) error) (restore func()) {
+	restore = testutil.Backup(&listenerRun)
+	listenerRun = f
+	return restore
+}
+
+func MockListenerReqs(f func(l *listener.Listener) <-chan *listener.Request) (restore func()) {
+	restore = testutil.Backup(&listenerRun)
```
Comment (bboozzoo) [inline overlord/ifacestate/apparmorprompting/export_test.go:42]: also ``s/listenerRun/listenerReqs/`

File: overlord/ifacestate/ifacemgr.go [hunk 25]
```diff
@@ -137,6 +138,20 @@ func Manager(s *state.State, hookManager *hookstate.HookManager, runner *state.T
 	return m, nil
 }
 
+// AppArmorPromptingRunning returns true if prompting is running.
```
Comment (pedronis) [inline overlord/ifacestate/ifacemgr.go:141]: this needs a not that it can only be called after the StartUp of ifacestate manager

File: overlord/ifacestate/ifacemgr.go [hunk 26]
```diff
@@ -174,6 +189,21 @@ func (m *InterfaceManager) StartUp() error {
 	if _, err := m.reloadConnections(""); err != nil {
 		return err
 	}
+
+	// check whether AppArmor prompting is supported, it is fine to do it
+	// once as toggling the feature imposes a restart of snapd
+	if assessAppArmorPrompting(m) {
+		m.useAppArmorPrompting = true
```
Comment (pedronis) [inline overlord/ifacestate/ifacemgr.go:196]: to be on the safe side I would move this bit very early here, just after we have the lock if possible or as soon as possible? and then reconsult the flag to run the manager

File: overlord/ifacestate/ifacemgr.go [hunk 27]
```diff
@@ -137,6 +138,20 @@ func Manager(s *state.State, hookManager *hookstate.HookManager, runner *state.T
...
+func (m *InterfaceManager) AppArmorPromptingRunning() bool {
+	return m.useAppArmorPrompting
+}
+
+// Allow m.UseAppArmorPrompting to be mocked in tests
+var assessAppArmorPrompting = func(m *InterfaceManager) bool {
+	return m.assesAppArmorPrompting()
+}
+
+func (m *InterfaceManager) InterfacesRequestsManager() *apparmorprompting.InterfacesRequestsManager {
```
Comment (pedronis) [inline overlord/ifacestate/ifacemgr.go:151]: this needs a doc comment and should also mention that it can be used only after StartUp

File: overlord/ifacestate/apparmorprompting/prompting.go [hunk 28]
```diff
@@ -0,0 +1,554 @@
...
+	defer m.lock.RUnlock()
+	return m.prompts.PromptWithID(userID, promptID)
+}
+
+// HandleReply checks that the given reply contents are valid, satisfies the
+// original request, and does not conflict with any existing rules (if lifespan
+// is not "single"). If all of these are true, sends a reply for the prompt with
+// the given ID, and both creates a new rule and checks any outstanding prompts
+// against it, if the lifespan is not "single".
+func (m *InterfacesRequestsManager) HandleReply(userID uint32, promptID prompting.IDType, constraints *prompting.Constraints, outcome prompting.OutcomeType, lifespan prompting.LifespanType, duration string) (satisfiedPromptIDs []prompting.IDType, retErr error) {
```
Comment (pedronis) [inline overlord/ifacestate/apparmorprompting/prompting.go:371]: thanks, this is now a bit easier to follow

File: overlord/ifacestate/apparmorprompting/prompting.go (hunk 18 above)
Comment (pedronis) [reply overlord/ifacestate/apparmorprompting/prompting.go:242]: would be good to clarify @bboozzoo question

**Review by pedronis — COMMENTED** [review]
apart the permission mgmt question this looks reasonable now, some final small comments

**Review by pedronis — APPROVED** [review]
thank you

**Review by bboozzoo — APPROVED** [review]
Thanks
